### PR TITLE
feat(ws): Add support for custom WebSocket

### PR DESCRIPTION
**Pull request by mglyzin**
## Ability to extend the original `WebSocket` class

This patch add type definitions for custom `WebSocket` class. 
This feature was merged in the original library at https://github.com/websockets/ws/pull/2007

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

### CODE REVIEW

**Review by vansergen — APPROVED** [review]

**Review by jlalmes — APPROVED** [review]

File: types/ws/ws-tests.ts [hunk 1]
```diff
@@ -314,3 +314,18 @@ function f() {
...
+            return 'foo';
+        }
+    }
+    const webSocketServer = new WebSocket.WebSocketServer<CustomWebSocket>({WebSocket: CustomWebSocket});
+    webSocketServer.on('connection', (ws) => {
+        // $ExpectType CustomWebSocket
+        ws;
+        // $ExpectType "foo"
+        ws.foo();
+    });
```
Comment (vansergen) [inline types/ws/ws-tests.ts:330]:
```suggestion
    });
    webSocketServer.clients.forEach((ws) => {
      // $ExpectType CustomWebSocket
      ws;
      // $ExpectType "foo"
      ws.foo();
    });
```

File: types/ws/ws-tests.ts [hunk 2]
```diff
@@ -328,7 +328,7 @@ function f() {
         // $ExpectType "foo"
         ws.foo();
     });
-    webSocketServer.clients.forEach((ws) => {
+    Array.from(webSocketServer.clients).forEach((ws) => {
```
Comment (vansergen) [inline types/ws/ws-tests.ts:331]: for future reference: 
```
ERROR: 332:1   expect                     TypeScript@4.6: Can not match a node to this assertion.
ERROR: 334:1   expect                     TypeScript@4.6: Can not match a node to this assertion.
```

File: types/ws/index.d.ts [hunk 3]
```diff
@@ -310,36 +311,36 @@ declare namespace WebSocket {
...
         once(event: "headers", cb: (this: Server, headers: string[], request: IncomingMessage) => void): this;
         once(event: "close" | "listening", cb: (this: Server) => void): this;
         once(event: string | symbol, listener: (...args: any[]) => void): this;
 
-        off(event: "connection", cb: (this: Server, socket: WebSocket, request: IncomingMessage) => void): this;
+        off(event: "connection", cb: (this: Server, socket: T, request: IncomingMessage) => void): this;
         off(event: "error", cb: (this: Server, error: Error) => void): this;
         off(event: "headers", cb: (this: Server, headers: string[], request: IncomingMessage) => void): this;
         off(event: "close" | "listening", cb: (this: Server) => void): this;
         off(event: string | symbol, listener: (this: Server, ...args: any[]) => void): this;
```
Comment (vansergen) [inline types/ws/index.d.ts:335]:
```suggestion
        off(event: "connection", cb: (this: Server<T>, socket: T, request: IncomingMessage) => void): this;
        off(event: "error", cb: (this: Server<T>, error: Error) => void): this;
        off(event: "headers", cb: (this: Server<T>, headers: string[], request: IncomingMessage) => void): this;
        off(event: "close" | "listening", cb: (this: Server<T>) => void): this;
        off(event: string | symbol, listener: (this: Server<T>, ...args: any[]) => void): this;
```

File: types/ws/index.d.ts [hunk 4]
```diff
@@ -310,36 +311,36 @@ declare namespace WebSocket {
...
         on(event: "headers", cb: (this: Server, headers: string[], request: IncomingMessage) => void): this;
         on(event: "close" | "listening", cb: (this: Server) => void): this;
         on(event: string | symbol, listener: (this: Server, ...args: any[]) => void): this;
 
-        once(event: "connection", cb: (this: Server, socket: WebSocket, request: IncomingMessage) => void): this;
+        once(event: "connection", cb: (this: Server, socket: T, request: IncomingMessage) => void): this;
         once(event: "error", cb: (this: Server, error: Error) => void): this;
         once(event: "headers", cb: (this: Server, headers: string[], request: IncomingMessage) => void): this;
         once(event: "close" | "listening", cb: (this: Server) => void): this;
         once(event: string | symbol, listener: (...args: any[]) => void): this;
```
Comment (vansergen) [inline types/ws/index.d.ts:329]:
```suggestion
        once(event: "connection", cb: (this: Server<T>, socket: T, request: IncomingMessage) => void): this;
        once(event: "error", cb: (this: Server<T>, error: Error) => void): this;
        once(event: "headers", cb: (this: Server<T>, headers: string[], request: IncomingMessage) => void): this;
        once(event: "close" | "listening", cb: (this: Server<T>) => void): this;
        once(event: string | symbol, listener: (this: Server<T>, ...args: any[]) => void): this;
```

File: types/ws/index.d.ts [hunk 5]
```diff
@@ -310,36 +311,36 @@ declare namespace WebSocket {
...
         ): void;
         shouldHandle(request: IncomingMessage): boolean | Promise<boolean>;
 
         // Events
-        on(event: "connection", cb: (this: Server, socket: WebSocket, request: IncomingMessage) => void): this;
+        on(event: "connection", cb: (this: Server<T>, socket: T, request: IncomingMessage) => void): this;
         on(event: "error", cb: (this: Server, error: Error) => void): this;
         on(event: "headers", cb: (this: Server, headers: string[], request: IncomingMessage) => void): this;
         on(event: "close" | "listening", cb: (this: Server) => void): this;
         on(event: string | symbol, listener: (this: Server, ...args: any[]) => void): this;
```
Comment (vansergen) [inline types/ws/index.d.ts:323]:
```suggestion
        on(event: "error", cb: (this: Server<T>, error: Error) => void): this;
        on(event: "headers", cb: (this: Server<T>, headers: string[], request: IncomingMessage) => void): this;
        on(event: "close" | "listening", cb: (this: Server<T>) => void): this;
        on(event: string | symbol, listener: (this: Server<T>, ...args: any[]) => void): this;
```

File: types/ws/index.d.ts [hunk 6]
```diff
@@ -310,36 +311,36 @@ declare namespace WebSocket {
...
+        on(event: "error", cb: (this: Server<T>, error: Error) => void): this;
+        on(event: "headers", cb: (this: Server<T>, headers: string[], request: IncomingMessage) => void): this;
+        on(event: "close" | "listening", cb: (this: Server<T>) => void): this;
+        on(event: string | symbol, listener: (this: Server<T>, ...args: any[]) => void): this;
+
+        once(event: "connection", cb: (this: Server<T>, socket: T, request: IncomingMessage) => void): this;
+        once(event: "error", cb: (this: Server<T>, error: Error) => void): this;
+        once(event: "headers", cb: (this: Server<T>, headers: string[], request: IncomingMessage) => void): this;
+        once(event: "close" | "listening", cb: (this: Server<T>) => void): this;
         once(event: string | symbol, listener: (...args: any[]) => void): this;
```
Comment (vansergen) [inline types/ws/index.d.ts:329]:
```suggestion
        once(event: string | symbol, listener: (this: Server<T>, ...args: any[]) => void): this;
```

**Review by vansergen — APPROVED** [review]

**Review by jlalmes — APPROVED** [review]

**Review by vansergen — APPROVED** [review]

**Comment by jlalmes** [comment]
LGTM, let's get this merged!

**Review by k-yle — CHANGES_REQUESTED** [review]
this has conflicts now that #58753 has been merged. Could you please rebase or merge master into this branch

File: types/ws/ws-tests.ts [hunk 7]
```diff
@@ -314,3 +314,34 @@ function f() {
...
+        // $ExpectType "foo"
+        ws.foo();
+    });
+    Array.from(webSocketServer.clients).forEach((ws) => {
+        // $ExpectType CustomWebSocket
+        ws;
+        // $ExpectType "foo"
+        ws.foo();
+    });
+    webSocketServer.on('upgrade', (request, socket, head) => {
```
Comment (vansergen) [inline types/ws/ws-tests.ts:337]:
```suggestion
    const server = new http.Server();
    server.on('upgrade', (request, socket, head) => {
```

**Review by vansergen — CHANGES_REQUESTED** [review]

**Review by vansergen — APPROVED** [review]
LGTM

**Review by jlalmes — APPROVED** [review]

**Review by vansergen — COMMENTED** [review]
nit ([#L1](https://github.com/mglyzin/DefinitelyTyped/blob/e9c0f22c872989e3dfc0b7184e6c8f1b702189bb/types/ws/index.d.ts#L1)): 

```typescript
// Type definitions for ws 8.5
```

Ref: [ws@8.5.0](https://github.com/websockets/ws/releases/tag/8.5.0)

**Review by k-yle — APPROVED** [review]

**Comment by jlalmes** [comment]
Hi all, this PR is now blocking us - what need to happen next so that we can merge & release this?

**Comment by k-yle** [comment]
@jlalmes since this is a critical package it's waiting in the queue for someone from Microsoft to review

**Comment by cwadrupldijjit** [comment]
I assume that also means it could be ready tomorrow, as today was a US
holiday.  When they review it, it'll go quick.

**Review by jakebailey — APPROVED** [review]

**Comment by mglyzin** [comment]
> 
Ready to merge

**Review by jakebailey — APPROVED** [review]

**Review by k-yle — APPROVED** [review]

File: types/ws/index.d.ts [hunk 8]
```diff
@@ -64,9 +64,9 @@ declare class WebSocket extends EventEmitter {
     onclose: (event: WebSocket.CloseEvent) => void;
     onmessage: (event: WebSocket.MessageEvent) => void;
```
Comment (vansergen) [inline types/ws/index.d.ts:66]:
```suggestion

     constructor(address: null);
```

**Review by vansergen — CHANGES_REQUESTED** [review]

File: types/ws/index.d.ts [hunk 9]
```diff
@@ -64,7 +64,8 @@ declare class WebSocket extends EventEmitter {
...
 
-    constructor(address: string | URL, options?: WebSocket.ClientOptions | ClientRequestArgs);
+    constructor(address: null);
+    constructor(address: string | null, options?: WebSocket.ClientOptions | ClientRequestArgs);
```
Comment (vansergen) [inline types/ws/index.d.ts:68]:
```suggestion
    constructor(address: string | URL, options?: WebSocket.ClientOptions | ClientRequestArgs);
```

**Review by vansergen — CHANGES_REQUESTED** [review]

**Review by jlalmes — APPROVED** [review]

**Review by vansergen — APPROVED** [review]

**Comment by mglyzin** [comment]
What's going on with this merge request? When it will be merged?

**Comment by jakebailey** [comment]
I was waiting to see if you'd address @vansergen's outstanding comments, but they approved the PR so I'll do so as well.

**Review by jakebailey — APPROVED** [review]

**Review by k-yle — APPROVED** [review]

**Comment by jakebailey** [comment]
Very weird that the bot re-pinged when nothing changed and it had already pinged. Oh well...

**Comment by k-yle** [comment]
Ready to merge

**Pull request merged**